File: src/truco/cards.py
"""Cards, ranks, suits, the 40-card Truco deck, and strength under a manilha rank."""
from __future__ import annotations
from dataclasses import dataclass

# Ranks low -> high: 4, 5, 6, 7, Q, J, K, A, 2, 3
RANK_NAMES = ["4", "5", "6", "7", "Q", "J", "K", "A", "2", "3"]
NUM_RANKS = 10

# Suit ordering for manilha tie-break (low -> high): diamond, spade, heart, club
SUIT_NAMES = ["D", "S", "H", "C"]
NUM_SUITS = 4


@dataclass(frozen=True, order=True)
class Card:
    rank: int   # 0..9
    suit: int   # 0..3

    def __str__(self) -> str:
        return f"{RANK_NAMES[self.rank]}{SUIT_NAMES[self.suit]}"

    def __repr__(self) -> str:
        return str(self)


ALL_CARDS: tuple[Card, ...] = tuple(
    Card(r, s) for r in range(NUM_RANKS) for s in range(NUM_SUITS)
)
assert len(ALL_CARDS) == 40


def parse_card(s: str) -> Card:
    s = s.strip().upper()
    if len(s) < 2:
        raise ValueError(f"Bad card: {s!r}")
    rank_s, suit_s = s[:-1], s[-1]
    if rank_s not in RANK_NAMES:
        raise ValueError(f"Bad rank: {rank_s!r}")
    if suit_s not in SUIT_NAMES:
        raise ValueError(f"Bad suit: {suit_s!r}")
    return Card(RANK_NAMES.index(rank_s), SUIT_NAMES.index(suit_s))


def manilha_rank(vira: Card) -> int:
    """Rank above the vira, wrapping 3 -> 4."""
    return (vira.rank + 1) % NUM_RANKS


def card_strength(card: Card, m_rank: int) -> int:
    """Higher = stronger. Manilhas dominate all non-manilhas; suit breaks ties among manilhas."""
    if card.rank == m_rank:
        return 100 + card.suit
    return card.rank

File: truco-rs/src/cards.rs
//! 40-card Truco deck. Ranks low->high: 4,5,6,7,Q,J,K,A,2,3. Suits D<S<H<C.

#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash, PartialOrd, Ord)]
pub struct Card {
    pub rank: u8, // 0..10
    pub suit: u8, // 0..4
}

pub const NUM_RANKS: u8 = 10;
pub const NUM_SUITS: u8 = 4;
pub const NUM_CARDS: usize = 40;

pub const RANK_NAMES: [&str; 10] = ["4", "5", "6", "7", "Q", "J", "K", "A", "2", "3"];
pub const SUIT_NAMES: [&str; 4] = ["D", "S", "H", "C"];

pub fn all_cards() -> [Card; NUM_CARDS] {
    let mut out = [Card { rank: 0, suit: 0 }; NUM_CARDS];
    let mut i = 0;
    for r in 0..NUM_RANKS {
        for s in 0..NUM_SUITS {
            out[i] = Card { rank: r, suit: s };
            i += 1;
        }
    }
    out
}

pub fn manilha_rank(vira: Card) -> u8 {
    (vira.rank + 1) % NUM_RANKS
}

pub fn card_strength(c: Card, m_rank: u8) -> u16 {
    if c.rank == m_rank {
        100u16 + c.suit as u16
    } else {
        c.rank as u16
    }
}

impl Card {
    pub fn fmt(&self) -> String {
        format!("{}{}", RANK_NAMES[self.rank as usize], SUIT_NAMES[self.suit as usize])
    }
}
